relayer/chains/bitcoin/provider.go

```diff
@@ -0,0 +1,1437 @@
+		client:             client,
+		LastSerialNumFunc:  func() *big.Int { return big.NewInt(0) },
+		httpServer:         make(chan struct{}),
+		db:                 db, // Add the database to the Provider
+		chainParam:         chainParam,
+		multisigAddrScript: msPubkey,
+	}
+	// Run an http server to help btc interact each others
+	go func() {
+		if c.Mode == MasterMode {
+			startMaster(c)
+		} else {
+			startSlave(c, p)
+		}
+		close(p.httpServer)
+	}()
+
+	return p, nil
+}
+
+func (p *Provider) CallSlaves(slaveRequestData []byte, path string) [][][]byte {
+	resultChan := make(chan [][][]byte)
+	go func() {
+		responses := make(chan slaveResponse, 2)
+		var wg sync.WaitGroup
+		wg.Add(2)
+
+		go requestPartialSign(p.cfg.ApiKey, p.cfg.SlaveServer1+path, slaveRequestData, responses, 1, &wg)
+		go requestPartialSign(p.cfg.ApiKey, p.cfg.SlaveServer2+path, slaveRequestData, responses, 2, &wg)
+
+		go func() {
+			wg.Wait()
+			close(responses)
+		}()
+
+		results := make([][][]byte, 2)
+		for res := range responses {
+			if res.err != nil {
+				p.logger.Error("failed to call slaves", zap.Error(res.err))
+				continue
+			}
+			results[res.order-1] = res.sigs
+		}
+		resultChan <- results
+	}()
+
+	return <-resultChan
+}
+
+func (p *Provider) QueryLatestHeight(ctx context.Context) (uint64, error) {
+	return p.client.GetLatestBlockHeight(ctx)
+}
+
+func (p *Provider) QueryTransactionReceipt(ctx context.Context, txHash string) (*relayTypes.Receipt, error) {
+	res, err := p.client.GetTransactionReceipt(ctx, txHash)
+	if err != nil {
+		return nil, err
+	}
+	return &relayTypes.Receipt{
+		TxHash: res.Txid,
+	}, nil
+}
+
+func (p *Provider) NID() string {
+	return p.cfg.NID
+}
+
+func (p *Provider) Name() string {
+	return p.cfg.ChainName
+}
+
+func (p *Provider) Init(ctx context.Context, homePath string, kms kms.KMS) error {
+	return nil
+}
+
+// Wallet returns the wallet of the provider
+func (p *Provider) Wallet() (*multisig.MultisigWallet, error) {
+	return p.buildMultisigWallet()
+}
+
+func (p *Provider) Type() string {
+	return p.cfg.ChainName
+}
+
+func (p *Provider) Config() provider.Config {
+	return p.cfg
+}
+
+func (p *Provider) Listener(ctx context.Context, lastProcessedTx relayTypes.LastProcessedTx, blockInfoChan chan *relayTypes.BlockInfo) error {
+	// run http server to help btc interact each others
+	latestHeight, err := p.QueryLatestHeight(ctx)
+	if err != nil {
+		p.logger.Error("failed to get latest block height", zap.Error(err))
+		return err
+	}
+	lastSavedHeight := lastProcessedTx.Height
+	startHeight, err := p.getStartHeight(latestHeight, lastSavedHeight)
+	if err != nil {
+		p.logger.Error("failed to determine start height", zap.Error(err))
+		return err
+	}
+
+	pollHeightTicker := time.NewTicker(time.Second * 60)
+	pollHeightTicker.Stop()
+
+	p.logger.Info("Start from height", zap.Uint64("height", startHeight), zap.Uint64("finality block", p.FinalityBlock(ctx)))
+
+	for {
+		select {
+		case <-pollHeightTicker.C:
```

Comment:
Add sleep or avoid using default so that the cpus usage would not be 100% within the loop.